### ISSUE TITLE
Getting the 255,000 peak dataset into usable shape

### ISSUE THREAD

**Issue by wkumler** [wkumler/MS_metrics#1]
In 2020, [Erik Müller](https://sciprofiles.com/profile/999203) et al. published a [dataset of 255,000 randomly selected and manually classified EICs for the evaluation of peak detection (and gap filling) methods](https://www.mdpi.com/2218-1989/10/4/162). This looks like the perfect data source for this project but the data's organized in a relatively unhelpful way as a [Zenodo repository](https://doi.org/10.5281/zenodo.3756211). There doesn't seem to be any code available online so we're going to be guessing at some of the steps, guided by the methods section provided in the paper.

The goal is to end up with a simple table with peak boundaries and the expert classification (good, bad, unknown). This will then allow us to extract the original data and can calculate all kinds of peak metrics on them, producing a useful data set:

|filename                   |  peak_mz| peak_rt_start| peak_rt_end|peak_quality |
|:--------------------------|--------:|-------------:|-----------:|:------------|
|161010_14_Stich_35_T.mzML  | 408.2227|      9.484196|    9.563940|Maybe peak   |
|161010_38_Stich_42.mzML    | 627.4868|     15.384681|   15.442389|Not peak     |
|161010_42_KA_HBS_1100.mzML | 320.1704|      7.128941|    7.455862|Maybe peak   |

**Comment by wkumler** [comment]
Some initial attempts revealed that the row IDs provided don't seem to match up. In the "Table_2_Holtemme_Stichtag2015_water_ESIpos_MRP2.7_gap-filled_clean.xlsx" file, there seems to be six columns appended to the front of the normal MzMine output with summary statistics for each row, but these summary statistics are different from the rest of the row.

In the first row, for example, the row's m/z is given as 105.0698656, but the entries in the MzMine table all have m/z's around 480.4254518. The retention times are also wildly different. When we went in and looked at the raw data, there was indeed a peak at 480.4254 that corresponded to the EIC image embedded in the classification Excel file (2_Classified_Data\Excel_tables_for_classification\peak_files\160923_53_Blank.mzML.xlsx)
 but only after we included zero values for each scan that was missing data.

We became even more suspicious when we did the joining/matching using the row IDs and filenames and learned that there were a bunch of peaks that were flagged as "good" peaks, but had no associated data. peak_mz, peak_rt_start, and peak_rt_end were all zero (which we think is what happens when MzMine does gap filling but finds no data points in the peak bounding box) and when we guessed at the proper m/z and retention time using the peaks that were initially detected we found no data there. This was the case for a very large number of peaks:

|           | Missing| Present|
|:----------|-------:|-------:|
|Maybe peak |   84318|  108173|
|Not peak   |   22294|   32038|
|Yes peak   |    3311|    4305|

where "Missing" means that the peak had zeroes in the peak_mz, peak_rt_start, and peak_rt_end columns.

**Comment by wkumler** [comment]
This is pretty catastrophic for the project unless we can find a way around it. Bryna suggested that we try joining the two tables together with a bind instead of joining on row ID because it seems like there's a one-to-one correspondence between the two, but the tables actually have different dimensions. The classification table (2_Classified_Data\Classification\Classification_after_cleanup.csv) has 5,000 rows, as expected, but the MzMine output table only has 4,989. The "missing" row IDs are

```2784 14477 14129 11509  4743  1896 26526 28737 17157 11484 13394```

and I have no idea how to deal with this, but it breaks the one-to-one row correspondence. I wonder if I can drop all the rows that are in this vector from the classification sheet and recover a one-to-one correspondence? Otherwise I'll need to connect row ID to the proper MzMine output using the average m/z and RT of each row.

**Comment by wkumler** [comment]
Nope, dropping those rows didn't seem to work. It did maybe improve things but it's still nowhere near trustworthy, with over a thousand "good" peaks having no associated data:

```
raw_classes <- read_csv(paste0("training data/A_dataset_for_evaluation_",
                               "of_peak_detection_methods_v3/",
                               "2_Classified_Data/Classification/",
                               "Classification_after_cleanup.csv"))

raw_peakinfo <- read_csv(paste0("training data/A_dataset_for_evaluation_",
                                "of_peak_detection_methods_v3/",
                "1_Processed_data/Selected_subset_MZmine_table/Table_2_",
                "Holtemme_Stichtag2015_water_ESIpos_MRP2.7_gap-filled_clean.csv"))
filename_order <- str_remove(str_subset(colnames(raw_peakinfo), "Peak m/z"), " Peak m/z")
peak_mzs <- raw_peakinfo %>%
  select(contains("Peak m/z")) %>%
  pivot_longer(cols = everything(), names_to = "filename", values_to = "peak_mz") %>%
  select(-filename)
peak_starts <- raw_peakinfo %>%
  select(contains("Peak RT start")) %>%
  pivot_longer(cols = everything(), names_to = "filename", values_to = "peak_rt_start") %>%
  select(-filename)
peak_ends <- raw_peakinfo %>%
  select(contains("Peak RT end")) %>%
  pivot_longer(cols = everything(), names_to = "filename", values_to = "peak_rt_end") %>%
  select(-filename)


clean_classes <- raw_classes %>%
  filter(!`MZmine row ID`%in%setdiff(`MZmine row ID`, raw_peakinfo$`row ID`)) %>%
  column_to_rownames(var = "MZmine row ID") %>% 
  select(all_of(filename_order)) %>%
  pivot_longer(cols = everything(), values_to = "peak_quality", names_to = "filename") %>%
  mutate(peak_quality=case_when(
    is.na(peak_quality) ~ "Yes peak",
    peak_quality == 1 ~ "Not peak",
    peak_quality == 0 ~ "Maybe peak"
  )) %>%
  cbind(peak_mzs) %>%
  cbind(peak_starts) %>%
  cbind(peak_ends)

clean_classes %>%
  mutate(missingness=ifelse(peak_mz==0, "Missing", "Present")) %>%
  with(table(peak_quality, missingness)) %>%
  knitr::kable()

```

|           | Missing| Present|
|:----------|-------:|-------:|
|Maybe peak |  100688|   91803|
|Not peak   |    7907|   46425|
|Yes peak   |    1328|    6288|

I think I'm only going to trust these numbers when I can get the Missing x Yes peak corner of this table down to less than 10(?). It's possible that some peaks were erroneously categorized during this analysis but I can't imagine that happening more than 1% of the time. I also want essentially all of the "Missing" numbers to be in the "Not peak" row because I can't imagine why you'd flag an empty chromatogram as needing more information (i.e. "Maybe peak").

On to the next option: aligning rows with row IDs using the mean m/z and RT of each row, calculated manually. Ugh.

**Comment by wkumler** [comment]
Okay, that method also didn't work. It didn't seem to produce any fewer missing "Yes" peaks and we lost a couple of peaks because we also couldn't disambiguate them.

**Comment by wkumler** [comment]
The new strategy may be to train up a dataset in a feature-based way - see #2

**Comment by wkumler** [comment]
Update: the initial data set of ALL MzMine output doesn't seem to be corrupted in the same way. The row summaries in the first six columns correspond well with the information in the MzMine table - the m/zs are similar, the rts make sense, and using Excel's COUNTIF(F2:XG2, "DETECTED") produces values that match the "row number of detected peaks" column. The problem is that I don't know which of these were randomly selected because I'm not sure whether the row IDs still match up.

![image](https://user-images.githubusercontent.com/43116923/214381925-1afd963f-f531-4775-a480-045f881a3965.png)

I wonder if I can use this uncorrupted data table, filter for the IDs that are in the classification Excel document, and skip the corrupted "clean" version this way?

**Comment by wkumler** [comment]
Answer: nope. The "original" data table seems to be somehow corrupted/incomplete as well. There are clear gaps in the table despite it (probably) initially having one row for every entry:

![image](https://user-images.githubusercontent.com/43116923/214390561-bcc64de0-2930-4af0-929f-149b86252fea.png)

and several rows that exist in the classification CSV don't exist in the original (eight, down from the ten I identified earlier: `1896  2784  4743 11509 14129 17157 26526 28737`)

![image](https://user-images.githubusercontent.com/43116923/214390934-f3eb890c-7252-4e6e-9c34-a62063fa497b.png)

**Comment by wkumler** [comment]
Some code for checking the above more systematically:

```r

library(tidyverse)
clean_peakclass <- read_csv(paste0("training data/A_dataset_for_evaluation_of_peak_detection_methods_v3/",
                                   "2_Classified_Data/Classification/",
                                   "Classification_after_cleanup.csv")) %>%
  pivot_longer(cols = -`MZmine row ID`) %>%
  select(row_id=`MZmine row ID`, filename=`name`, peak_quality=`value`) %>%
  mutate(peak_quality=case_when(
    is.na(peak_quality) ~ "Maybe peak",
    peak_quality == 1 ~ "Yes peak",
    peak_quality == 0 ~ "Not peak"
  )) %>%
  arrange(filename, row_id)


peakinfo_by_file <- read_csv(paste0("training data/A_dataset_for_evaluation_of_peak_detection_methods_v3/1_Processed_data/Original_MZmine_table/Table_1_Holtemme_Stichtag2015_water_ESIpos_MRP2.7_gap-filled.csv"))
clean_peakinfo <- peakinfo_by_file %>%
  filter(`row ID`%in%unique(clean_peakclass$row_id)) %>%
  select(row_id=`row ID`, row_mz=`row m/z`, row_rt=`row retention time`, 
         contains("Peak m/z"), contains("Peak RT start"), contains("Peak RT end")) %>%
  pivot_longer(cols = -c(row_id, row_mz, row_rt)) %>%
  separate("name", into = c("filename", "metric"), sep = " Peak ") %>%
  pivot_wider(names_from = "metric", values_from = "value") %>%
  rename(peak_mz=`m/z`, peak_rt_start=`RT start`, peak_rt_end=`RT end`) %>%
  arrange(filename, row_id)

peak_data <- clean_peakinfo %>%
  left_join(clean_peakclass) %>%
  select(row_id, filename, peak_mz, peak_rt_start, peak_rt_end, peak_quality)

peak_data %>%
  mutate(missingness=ifelse(peak_mz==0, "Missing", "Present")) %>%
  with(table(peak_quality, missingness))
```

**Comment by wkumler** [comment]
The author of the paper reached out to me in response to my email question and has been responsive to the concerns I've raised. In particular, he clarified that there is indeed an issue with the Zenodo dataset and provided access to a new version of the "complete" (not 5,000 subset) MzMine output. However, he also confirmed that the original descriptors are correct (1 = no peak, 0 = ambiguous, and empty/NA = good peak). I remain pretty skeptical of this for the reasons detailed above.

Fortunately, with the new dataset I was able to go back and recalculate my peak quality metrics for the 255,000 peaks involved. They appear to confirm my suspicions - that the proper encoding is 1=Good, 0=Bad, and empty = Ambiguous. With my peak metrics, I get higher SNR and peak_cor values for the peaks encoded as 1s than I do the NAs than I do the 0s:

![image](https://user-images.githubusercontent.com/43116923/222820959-c6eaa255-4f62-4ad9-ae55-8eb0f9c9cfee.png)

![image](https://user-images.githubusercontent.com/43116923/222821295-1a43d5d8-a727-4bd0-9170-6e87052067a1.png)

I'm a little reluctant to move forward with this dataset while this is still unresolved, but I can't tell whether I'm being obnoxious and insistent that my metrics are working when I should be downgrading them or if there's truly still issues with the dataset itself.
